amads/melody/contour/interpolation_contour.py

```diff
@@ -0,0 +1,272 @@
+
+    def __init__(self, times: list[float], pitches: list[int]):
+        """Initialize with time and pitch values.
+
+        Parameters
+        ----------
+        times : list[float]
+            Array of onset times in seconds
+        pitches : list[int]
+            Array of pitch values
+
+        Raises
+        ------
+        ValueError
+            If times and pitches are not the same length
+
+        Examples
+        --------
+        >>> ic = InterpolationContour([0, 0.1, 0.2, 0.3], [60, 58, 64, 65])
+        >>> len(ic.contour)
+        30
+
+        >>> ic = InterpolationContour([0, 0.1, 0.2, 0.3], [60, 58, 64, 65, 65])
+        Traceback (most recent call last):
+            ...
+        ValueError: Times and pitches must have the same length
+        """
+        if len(times) != len(pitches):
+            raise ValueError("Times and pitches must have the same length")
+        self.times = times
+        self.pitches = pitches
+        self.contour = self.calculate_interpolation_contour()
+
+    @staticmethod
+    def _find_contour_extrema(pitches: list[int]) -> list[int]:
+        """Determine contour extremum notes (local minima and maxima, including endpoints).
+
+        Excludes changing notes (notae cambiatae).
+
+        Parameters
+        ----------
+        pitches : list[int]
+            List of MIDI pitch values
+
+        Returns
+        -------
+        numpy.ndarray
+            Array of indices corresponding to extrema points in the melody
+        """
+        extrema_indices = []
+        n = len(pitches)
+
+        # Always include first and last notes
+        extrema_indices.append(0)
+
+        # Find local extrema
+        for i in range(1, n - 1):
+            # Check previous and next notes
+            prev_pitch = pitches[i - 1]
+            curr_pitch = pitches[i]
+            next_pitch = pitches[i + 1]
+
+            # Skip changing notes (where adjacent pitches are equal)
+            if prev_pitch == next_pitch:
+                continue
+
+            # Check if it's a peak or valley
+            if (prev_pitch < curr_pitch and next_pitch < curr_pitch) or (
+                prev_pitch > curr_pitch and next_pitch > curr_pitch
+            ):
+                # Additional check for equal adjacent notes
+                is_extremum = True
+                if i > 1 and prev_pitch == curr_pitch:
+                    is_extremum = (pitches[i - 2] - curr_pitch) * (
+                        next_pitch - curr_pitch
+                    ) < 0
+                if i < n - 2 and next_pitch == curr_pitch:
+                    is_extremum = (prev_pitch - curr_pitch) * (
+                        pitches[i + 2] - curr_pitch
+                    ) < 0
+                if is_extremum:
+                    extrema_indices.append(i)
+
+        extrema_indices.append(n - 1)  # Add last note
+        return np.array(extrema_indices)
+
+    def calculate_interpolation_contour(self) -> list[float]:
```

Comment:
I think it's redundant to have a method called `calculate_interpolation_contour`, I would just put this all into the `__init__` function? Up to you though

Reply:
I've updated the implementation after testing robustly against the original implementation - think I'd prefer to keep in a seperate method since it's ended up quite complex for inclusion in `__init__`